crates/core/tedge_agent/src/file_transfer_server/actor.rs

```diff
@@ -76,91 +92,250 @@ impl Builder<FileTransferServerActor> for FileTransferServerBuilder {
-            .with_data_dir(ttd.utf8_path_buf().into())
-            .with_port(port_in_use);
+        ensure!(
+            binding_res.is_err(),
+            "expected port binding to fail, but `try_bind` finished successfully"
+        );
 
-        let server = FileTransferServerBuilder::new(http_config).build().run();
+        Ok(())
+    }
 
-        tokio::select! {
-            res = server => ensure!(res.is_err(), "expected server startup to fail with port binding error, but actor exited successfully"),
-            _ = tokio::time::sleep(Duration::from_secs(5)) => bail!("timed out waiting for actor to stop running"),
-        }
+    #[tokio::test]
+    async fn server_uses_https_if_certificate_is_configured() -> anyhow::Result<()> {
+        let server_cert = rcgen::generate_simple_self_signed(["localhost".into()])
+            .context("generating server certificate")?;
+        let server = Server::new_https(server_cert, None).await?;
+        let test_url = server.url_for("test-file");
+
+        let client = server.anonymous_client()?;
+        let upload_response = client.put(&test_url).body("file").send().await.unwrap();
+        assert_eq!(upload_response.status(), hyper::StatusCode::CREATED);
 
         Ok(())
     }
+
+    #[tokio::test]
+    async fn server_accepts_connections_with_trusted_root_certificates() -> anyhow::Result<()> {
+        let server_cert = rcgen::generate_simple_self_signed(["localhost".into()])
+            .context("generating server certificate")?;
+        let client_cert = rcgen::generate_simple_self_signed(["a-client".into()])
+            .context("generating client certificate")?;
+        let server = Server::new_https(server_cert, Some(&client_cert)).await?;
+        let test_url = server.url_for("test-file");
+
+        let client = server.client_with_certificate(&client_cert)?;
+        let upload_response = client.put(&test_url).body("file").send().await.unwrap();
+        assert_eq!(upload_response.status(), hyper::StatusCode::CREATED);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn server_rejects_unauthenticated_connections_if_configured() -> anyhow::Result<()> {
+        let server_cert = rcgen::generate_simple_self_signed(["localhost".into()])
+            .context("generating server certificate")?;
+        let client_cert = rcgen::generate_simple_self_signed(["a-client".into()])
+            .context("generating client certificate")?;
+        let server = Server::new_https(server_cert, Some(&client_cert)).await?;
+
+        let client = server.anonymous_client()?;
+        let test_url = server.url_for("test/file");
+
+        let upload_err = client.put(&test_url).body("file").send().await.unwrap_err();
+        axum_tls::assert_error_matches(&upload_err, rustls::AlertDescription::CertificateRequired);
+
+        Ok(())
+    }
+
+    struct Server<Cert> {
+        port: u16,
+        temp_dir: TempTedgeDir,
+        server_cert: Cert,
+    }
+
+    impl Server<()> {
```

Comment:
question: Why do we define a separate server for testing reasons instead of testing the actual one? I think adding a comment would make it easier to understand for potential readers.

Reply:
"Server" is maybe not quite the right term here. I wanted some sort of abstraction to allow the tests to talk in terms of only the things that are relevant here, otherwise it's essentially impossible to see the subtle details around whether HTTPS is enabled, whether we expect client certs, whether we provide client certs in the request etc...

Reply:
I don't think this is particularly easy/valuable to generalise to other services, so I've instead opted for renaming this to `TestFileTransferService` with a short doc comment justifying its existence. I don't love having lots of ad-hoc utility code to simplify tests, but in this case, without this code it's very difficult to identify what each test is doing differently to the next.